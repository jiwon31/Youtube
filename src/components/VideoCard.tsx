import { useNavigate } from "react-router-dom";
import { VideoMetadata } from "types/video-type";
import { formatDate } from "util/formatDate";

type VideoCardProps = {
  video: VideoMetadata;
  type: string;
};

export default function VideoCard({ video, type }: VideoCardProps) {
  const { title, thumbnails, channelTitle, publishedAt } = video.snippet;
  const navigate = useNavigate();
  const isList = type === "list";

  return (
    <li
      className={`${
        isList ? "flex gap-2 m-2 ml-3" : ""
      } hover:scale-105 ease-in duration-200 cursor-pointer`}
      onClick={() => {
        navigate(`/videos/watch/${video.id}`, { state: { video } });
      }}
    >
      <img
        className={isList ? "w-40" : "w-full"}
        src={thumbnails.medium.url}
        alt={title}
      />
      <div>
        <p className="text-sm font-semibold my-2 line-clamp-2">{title}</p>
        <p className="text-xs opacity-80">{channelTitle}</p>
        <p className="text-xs opacity-80">{formatDate(publishedAt)}</p>
      </div>
    </li>
  );
}
