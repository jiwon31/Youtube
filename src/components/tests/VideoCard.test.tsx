import VideoCard from "components/VideoCard";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import { Route, useLocation } from "react-router-dom";
import { fakeVideo as video } from "tests/videos";
import { withRouter } from "tests/utils";
import { formatDate } from "util/formatDate";
import userEvent from "@testing-library/user-event";

describe("VideoCard", () => {
  const { title, thumbnails, channelTitle, publishedAt } = video.snippet;

  it("renders grid type correctly", () => {
    const component = renderer.create(
      withRouter(
        <Route path="/" element={<VideoCard video={video} type="" />}></Route>
      )
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("renders list type correctly", () => {
    const component = renderer.create(
      withRouter(
        <Route
          path="/"
          element={<VideoCard video={video} type="list" />}
        ></Route>
      )
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  // 정적 테스트를 스냅샷 테스트만 할건지 아니면 밑의 테스트도 할건지는 선택사항!
  it("renders video item", () => {
    render(
      withRouter(
        <Route path="/" element={<VideoCard video={video} type="" />}></Route>
      )
    );

    const image: HTMLImageElement = screen.getByRole("img");
    expect(image.src).toBe(thumbnails.medium.url);
    expect(image.alt).toBe(title);
    expect(screen.getByText(title)).toBeInTheDocument();
    expect(screen.getByText(channelTitle)).toBeInTheDocument();
    expect(screen.getByText(formatDate(publishedAt))).toBeInTheDocument();
  });

  it("navigate to detailed video page with video state when clicked", () => {
    function LocationStateDisplay() {
      return <pre>{JSON.stringify(useLocation().state)}</pre>;
    }

    render(
      withRouter(
        <>
          <Route path="/" element={<VideoCard video={video} type="" />}></Route>
          <Route
            path={`/videos/watch/${video.id}`}
            element={<LocationStateDisplay />}
          ></Route>
        </>
      )
    );

    const card = screen.getByRole("listitem");
    userEvent.click(card);

    expect(screen.getByText(JSON.stringify({ video }))).toBeInTheDocument();
  });
});
