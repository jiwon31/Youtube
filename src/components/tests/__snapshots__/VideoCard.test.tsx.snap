// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoCard renders grid type correctly 1`] = `
<li
  className=" hover:scale-105 ease-in duration-200 cursor-pointer"
  onClick={[Function]}
>
  <img
    alt="title"
    className="w-full"
    src="http://image/"
  />
  <div>
    <p
      className="text-sm font-semibold my-2 line-clamp-2"
    >
      title
    </p>
    <p
      className="text-xs opacity-80"
    >
      channelTitle
    </p>
    <p
      className="text-xs opacity-80"
    >
      3 weeks ago
    </p>
  </div>
</li>
`;

exports[`VideoCard renders list type correctly 1`] = `
<li
  className="flex gap-2 m-2 ml-3 hover:scale-105 ease-in duration-200 cursor-pointer"
  onClick={[Function]}
>
  <img
    alt="title"
    className="w-40"
    src="http://image/"
  />
  <div>
    <p
      className="text-sm font-semibold my-2 line-clamp-2"
    >
      title
    </p>
    <p
      className="text-xs opacity-80"
    >
      channelTitle
    </p>
    <p
      className="text-xs opacity-80"
    >
      3 weeks ago
    </p>
  </div>
</li>
`;
